File: 2022/day-15/day15.py

import re
from multiprocessing import Pool, cpu_count
from typing import *

Point = Beacon = Sensor = Tuple[int, int]
Grid = List[Tuple[Beacon, Sensor]]

def parse_input(inp: str) -> Grid:
    coord_regex = re.compile(r'[x|y]=([+-]?\d+)')
    def parse_line(l: str) -> Tuple[Sensor, Beacon]:
        srow, scol, brow, bcol = map(int, coord_regex.findall(l))
        return (srow, scol), (brow, bcol)

    return list(map(parse_line, inp.splitlines()))

def dist(a: Point, b: Point) -> int:
    ar, ac = a
    br, bc = b
    return abs(br - ar) + abs(bc - ac)


def intervals_for_row(grid: Grid, target: int = 2000000) -> List[Point]:
    free_ranges = set()
    for sensor, nearest_beacon in grid:
        nearest_distance = dist(sensor, nearest_beacon)
        srow, scol = sensor
        dcol = abs(scol - target)
        if dcol <= nearest_distance:
            drow = nearest_distance - dcol
            free_ranges.add((srow - drow, srow + drow))

    first_range, *remaining_ranges = sorted(free_ranges)

    merged_intervals: List[Point] = [first_range]
    for rem_row, rem_col in remaining_ranges:
        last_row, last_col = merged_intervals[-1]
        if rem_row > last_col:
            merged_intervals.append((rem_row, rem_col))
        else:
            merged_intervals[-1] = (last_row, max(last_col, rem_col))

    return merged_intervals

def chunks(rmin: int, rmax: int, chunk_size: Optional[int] = None) -> Iterable[Iterable[int]]:
    chunk_size = chunk_size or abs(rmax - rmin) // cpu_count()
    if abs(rmax - rmin) % chunk_size != 0:
        raise ValueError
    indices = list(range(rmin, rmax+1, chunk_size))
    return [range(*r) for r in zip(indices[:-1], indices[1:])]



def find_freq(grid: Grid, rows: Iterable[int], max_row: int) -> Optional[int]:
    for row in rows:
        intervals = intervals_for_row(grid, row)
        if len(intervals) > 1:
            _, col = intervals[0]
            return (col +1)* max_row + row
    return None


def part_two(grid: Grid, max_row: int = 4000000):
    result = None
    with Pool() as pool:
        inputs = ((grid, rows, max_row) for rows in chunks(0, max_row))
        for freq in pool.starmap(find_freq, inputs, chunksize=1):
            if freq is not None:
                result = freq
    print('Part 2:', result)


def part_one(grid: Grid, target: int = 2000000):
    merge_col, merge_row = next(iter(intervals_for_row(grid, target)))
    print('Part 1:',  merge_row - merge_col)

with open('input.txt') as h:
    contents = h.read()

board = parse_input(contents)
part_one(board)
part_two(board)

File: 2022/day-15/day15.rs
use regex::Regex;
use std::cmp::max;

type Point = (i64, i64);
type Grid = Vec<(Point, Point)>;

fn parse_input(inp: &str) -> Grid {
    let coord_regex = Regex::new(r"([+-]?\d+)").unwrap();
    inp.lines()
        .map(|l| {
            let matches: Vec<i64> = coord_regex
                .find_iter(l)
                .map(|m| l[m.start()..m.end()].parse::<i64>().unwrap())
                .collect();
            assert_eq!(matches.len(), 4);
            ((matches[0], matches[1]), (matches[2], matches[3]))
        })
        .collect()
}

fn dist(a: &Point, b: &Point) -> i64 {
    (b.0 - a.0).abs() + (b.1 - a.1).abs()
}

fn intervals_for_row(grid: &Grid, target: i64) -> Vec<Point> {
    let mut free_ranges: Vec<Point> = Vec::new();
    for (sensor, nearest_beacon) in grid {
        let nearest_distance = dist(sensor, nearest_beacon);
        let (srow, scol) = *sensor;
        let dcol = (scol - target).abs();
        if dcol <= nearest_distance {
            let drow = nearest_distance - dcol;
            free_ranges.push((srow - drow, srow + drow))
        }
    }
    free_ranges.sort();
    let mut merged_intervals: Vec<Point> = Vec::from_iter(free_ranges.iter().take(1).cloned());
    for (rem_row, rem_col) in free_ranges.iter().skip(1).cloned() {
        let (last_row, last_col) = *merged_intervals.last().unwrap();
        if rem_row > last_col {
            merged_intervals.push((rem_row, rem_col));
        } else {
            *merged_intervals.last_mut().unwrap() = (last_row, max(last_col, rem_col));
        }
    }
    merged_intervals
}

fn part_one(grid: &Grid) {
    let intervals = intervals_for_row(grid, 2000000);
    let (merged_col, merged_row) = intervals.first().unwrap();
    println!("Part 1: {:?}", merged_row - merged_col);
}

fn part_two(grid: &Grid) {
    use rayon::prelude::*;

    let max_row: i64 = 4000000;
    let result: Vec<_> = (0..=max_row)
        .rev()
        .par_bridge()
        .filter_map(|row| {
            let intervals = intervals_for_row(grid, row);
            match intervals.len() {
                0..=1 => None,
                _ => {
                    let (_, col) = *intervals.first().unwrap();
                    Some((col + 1) * max_row + row)
                }
            }
        })
        .collect();
    println!("Part 2: {:?}", result.first().unwrap());
}

fn main() {
    let contents = std::fs::read_to_string("input.txt").expect("Failed to read input!");
    let grid = parse_input(&contents);
    part_one(&grid);
    part_two(&grid);
}
